File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode frontPointer;
    public boolean isPalindrome(ListNode head) {
        frontPointer = head;
        return backtrack(head);
    }
    public boolean backtrack(ListNode cur){
        if(cur!=null){
            if(!backtrack(cur.next)) return false;
            if(frontPointer.val != cur.val) return false;
            frontPointer=frontPointer.next;
        }
        return true;
    }
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* slow=head;
        ListNode* fast=head;
        while(fast!=NULL && fast->next!=NULL){
            slow=slow->next;
            fast=fast->next->next;
        }
        slow=reverse(slow,NULL);
        while(slow!=NULL){
            if(slow->val !=head->val){
                return false;
            }
            slow=slow->next;
            head=head->next;
        }
        return true;
    }
    ListNode* reverse(ListNode* cur,ListNode* prev){
        if(cur==NULL){
            return prev;
        }
        ListNode* t=cur->next;
        cur->next = prev;
        return reverse(t, cur);
    }
};